lib/wos_records.rb

```diff
@@ -0,0 +1,128 @@
+require 'forwardable'
+require 'htmlentities'
+
+# Utilities for working with Web of Science records
+class WosRecords
+  extend Forwardable
+  include Enumerable
+
+  # @return [Integer] WOS record count
+  def_delegators :rec_nodes, :count
+
+  # @return xml [String] WOS records in XML
+  def_delegators :doc, :to_xml
+
+  # @return doc [Nokogiri::XML::Document] WOS records document
+  attr_reader :doc
+
+  # @param records [String] records in XML
+  # @param encoded_records [String] records in HTML encoding
+  def initialize(records: nil, encoded_records: nil)
+    @records = records
+    @encoded_records = encoded_records
+    @records = decode_records if records.nil? && !encoded_records.nil?
+    @doc = Nokogiri::XML(@records) { |config| config.strict.noblanks }
+  end
+
+  # Iterate over the REC nodes
+  # @yield rec [Nokogiri::XML::Element]
+  def each
+    rec_nodes.each { |rec| yield rec }
+  end
+
+  # @return uids [Array<String>] the rec_nodes UID values (in order)
+  def uids
+    uid_nodes.map(&:text)
+  end
+
+  # @return uid_nodes [Nokogiri::XML::NodeSet] the rec_nodes UID nodes
+  def uid_nodes
+    rec_nodes.search('UID')
+  end
+
+  # Find duplicate WoS UID values
+  # @param record_setB [WosRecords]
+  # @return [Set] duplicate WoS UID values
+  def duplicate_uids(record_setB)
+    uids.to_set.intersection(record_setB.uids.to_set)
+  end
+
+  # Find duplicate WoS records
+  # @param record_setB [WosRecords]
+  # @return [Nokogiri::XML::NodeSet] duplicate records
+  def duplicate_records(record_setB)
+    duplicates = []
+    uid_intersection = duplicate_uids(record_setB)
+    unless uid_intersection.empty?
+      # create a new set of records, use a philosophy of immutability
+      # Nokogiri::XML::NodeSet enumerable methods do not return new objects
+      # Nokogiri::XML::Document.dup is a deep copy
```

Comment:
Note: the hangup with Nokogiri is unsurprising and doesn't require a bunch of commentary.  If you want to build a nodeset from existing nodes without retaining any connection back to the parent document (which all nodes have), then you can round-trip the node through serialization `.to_xml` and back into a new Nokogiri XML node/fragment.  This would reduce the memory utilization by not copying the whole document.

Reply:
Also, it is not a bug.

Reply:
We can discuss this in person, if necessary, as I like it as it is.  I hope it's not a blocker.  The round trip via `.to_xml` will not gain anything, as the doc only contains records anyway.  The explanation fills a gap in the Nokogiri documentation and I believe will serve as a note to prevent later refactoring that could easily mistakenly introduce violations of the immutability desired (which should be caught by specs, but might not be understood without this commentary).

Reply:
The round trip is what divorces it from the parent document (since it becomes just a String value, with no ability to point to location of the originating document in Ruby memory).  The separation is the basis of your motivation to `.dup`.  I agree this may benefit from better docs, but it is not unexpected behavior.

Reply:
I get how the `.to_xml` works.  What I'm saying is that the doc just contains REC elements, so there's nothing else duped by the `doc.dup` anyway, so there's not much gained from all the extra serialization transforms involved in the `.to_xml`; it will just slow things down and that might be worse that a faster `doc.dup` when collecting 100's of records.

Reply:
I agree there may be a computational tradeoff for the benefit of smaller memory footprint.  We can defer on this one for later.